Wrap the Single Product Template in a div with the product class

gigitux opened the pull request:
<!-- Start by describing the changes made in this Pull Request, and the reason for such changes. -->

This is an attempt to fix #8314. This PR is an exploration that I did, and I would like to receive feedback to see if it makes sense to wrap up my work.

<!-- Reference any related issues or PRs here -->

Fixes #8314

### Screenshots

### Testing

#### Automated Tests
* [x] Changes in this PR are covered by Automated Tests.
  * [x] Unit tests
  * [ ] E2E tests

#### User Facing Testing

<!-- Write these steps from the perspective of a "user" (merchant) familiar with WooCommerce. No need to spell out the steps for common setup scenarios (eg. "Create a product"), but do be specific about the thing being tested. Include screenshots demonstrating expectations where that will be helpful. -->

Check out this branch.

1. Edit the Single Product Template.
2. Be sure that a group block wraps the Single Product Template.
3. On the frontend page, be sure that the Single Product Template is wrapped with a div with the classes `woocommerce` and `product`.
4. Merge #8235 in this branch.
5. Follow the testing instructions of #8235.
6. You can notice that the style of the Product Image Gallery is loaded correctly.

* [x] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [ ] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

### Performance Impact

<!-- Please document any known performance impact (positive or negative) here. If negative, provide some rationale for why this is an okay tradeoff or how this will be addressed. -->

gigitux committed:
wrap the Single Product Template in a div with the product class

Aljullu commented on src/Templates/BlockTemplatesCompatibility.php:
```diff
@@ -361,4 +361,68 @@ protected function inject_attribute( &$block ) {
+
+		$new_block = parse_blocks(
+			sprintf(
+				'<!-- wp:group {"className":"woocommerce product"} -->
+				<div class="wp-block-group woocommerce product">
+					%1$s
+				</div>
+			<!-- /wp:group -->',
```
I don't have a strong opinion, but I'm curious to know why you defined a Group block instead of simply a `div`?

gigitux replied:
I'm using `parse_blocks` function to generate the new structure. In the WordPress template markup, the div is represented with a block `group`. This is the reason I decided to define a Group block.

Aljullu reviewed (commented):
This solution looks promising. Good job, @gigitux!

> As I said above, I didn't check all the cases. For example, my PR doesn't work if a group block doesn't wrap up the Single Product Template. For this reason, it is necessary to increase the unit test coverage and catch all the edge cases.

To solve this, I wonder if instead of wrapping the first block which is not inside a template part, maybe we should:

* Wrap **all blocks** starting from the first one which is not inside a template part and finishing at the last block before the next template part (aka footer) or the last block in the page in case there is no footer. In other words, wrapping all blocks between the header and the footer instead of only wrapping the first one. That would solve the issue you described, no?
* I would also add a check that inside the blocks we are wrapping, there is the _Add to Cart form_ block or another block we know is required in the Single Product page. That would solve the issue of adding the wrapper when not needed (ie, when using the WooCommerce Classic Template).

Do you think that would work?

gigitux commented:
Thanks for the review!

> * Wrap **all blocks** starting from the first one which is not inside a template part and finishing at the last block before the next template part (aka footer) or the last block in the page in case there is no footer. In other words, wrapping all blocks between the header and the footer instead of only wrapping the first one. That would solve the issue you described, no?

Yeah, good point! I updated the code :) 

> I would also add a check that inside the blocks we are wrapping, there is the Add to Cart form block or another block we know is required in the Single Product page. That would solve the issue of adding the wrapper when not needed (ie, when using the WooCommerce Classic Template).

👍

gigitux committed:
improve logic and increase coverage of unit test

gigitux committed:
improve logic and increase coverage of unit test

gigitux commented:
Hey @Aljullu, the PR is ready to review. 
I improved the logic and increased the coverage of unit test.

gigitux committed:
Merge branch 'trunk' into fix/8314-blockifying-single-product-template-investigate-how-to-apply-wc-core-styles-to-product-template

gigitux committed:
fix test

gigitux commented on src/Templates/BlockTemplatesCompatibility.php:
```diff
@@ -361,4 +361,126 @@ protected function inject_attribute( &$block ) {
+			}
+		);
+
+		$single_product_template_blocks = array( 'woocommerce/breadcrumbs', 'woocommerce/product-gallery-image', 'woocommerce/product-add-to-cart', 'woocommerce/product-details', 'woocommerce/add-to-cart-form' );
```
I would like to not hardcoded these slugs, but the name of the blocks isn't declared as constant :'(

Aljullu replied:
Agree, but I think it's ok for now.

Btw, why defining this variable here instead of inside `has_single_product_template_blocks()`?

gigitux replied:
Because it is a recursive function, I avoided creating multiple instances of the same array.

Aljullu replied:
Aaah, makes sense!

gigitux committed:
format HTML

Aljullu reviewed (commented):
This looks more robust, but there is still the (edge) case that if somebody adds a block above the _Header_, this will break:

Editor | Frontend
--- | ---
![imatge](https://user-images.githubusercontent.com/3616980/216609916-ed88d250-4fcc-450d-aa44-95a30f34663f.png) | ![imatge](https://user-images.githubusercontent.com/3616980/216610014-a46f43af-a1b7-4037-8186-037303a70939.png)

(Notice how the paragraph _This is a product_ and the Breadcrumb block are below the Header in the editor but above it in the frontend)

---

This is not a straight-forward issue because there are many possible edge cases. What I would try to do:

- Step 1: Split all blocks/template parts. Grouping blocks that don't have template parts in between.
- Step 2: Go through each block/group of blocks and run `has_single_product_template_blocks()`. If true, then run `create_wrap_block_group()`.

### Example

Imagine we have a page with a block above the header and a block below the footer:
* B1 + TP1 + B2 + B3 + B4 + TP2 + B5
_(B = Block, TP = Template Part)_

Step 1: You would split the blocks/template parts and store them in an array, using TP as delimiters:

* [[B1], [TP1], [B2, B3, B4], [TP2], [B5]]

Step 2: for each item of this array, you would just need to run `has_single_product_template_blocks()` and `create_wrap_block_group()` as needed.

If B2, B3 or B4 were a "product template block", you would end up with a wrapper between B2 and B4.

There might be other solutions, but this was the first one that came to my mind. I might have missed some other edge case as well.

gigitux commented:
> This looks more robust, but there is still the (edge) case that if somebody adds a block above the Header, this will break:

Good point. The assumption that I made is wrong. 👍 

> This is not a straight-forward issue because there are many possible edge cases. What I would try to do:
Step 1: Split all blocks/template parts. Grouping blocks that don't have template parts in between.
Step 2: Go through each block/group of blocks and run has_single_product_template_blocks(). If true, then run create_wrap_block_group().

Having multiple div with the same class could break some extensions, no? 🤔

Aljullu commented:
> Having multiple div with the same class could break some extensions, no? 🤔 

Hmmm, maybe, but some extensions could break as well if some "product template blocks" (ie: Product Gallery and Add to Cart) are not wrapped in `<div class="product">`, no? I would expect having multiple `<div class="product">` to be less problematic, but I might be wrong.

gigitux committed:
fix edge case

gigitux commented:
> > Having multiple div with the same class could break some extensions, no? 🤔
> 
> Hmmm, maybe, but some extensions could break as well if some "product template blocks" (ie: Product Gallery and Add to Cart) are not wrapped in `<div class="product">`, no? I would expect having multiple `<div class="product">` to be less problematic, but I might be wrong.

Yeah, I guess that this is the only way to go. I updated the PR and added another test.

Aljullu commented on src/Templates/BlockTemplatesCompatibility.php:
```diff
@@ -369,31 +369,42 @@ protected function inject_attribute( &$block ) {
-		}
+		$wrapped_blocks = array_map(
+			function( $blocks ) use ( $single_product_template_blocks ) {
+				$has_single_product_template_blocks = false;
```
Can this line be removed? In L385 we are setting `$has_single_product_template_blocks` again.

Aljullu commented on src/Templates/BlockTemplatesCompatibility.php:
```diff
@@ -448,38 +459,35 @@ private static function has_single_product_template_blocks( $parsed_blocks, $sin
 					return $carry;
 				}
-				array_push( $carry['blocks'], $block );
+				$last_element_index = count( $carry ) - 1;
+				if ( isset( $carry[ $last_element_index ][0]['blockName'] ) && 'core/template-part' !== $carry[ $last_element_index ][0]['blockName'] ) {
+					array_push( $carry[ $last_element_index ], $block );
+					return $carry;
+				}
```
I guess you can use the [`end()` PHP function](https://www.php.net/end) here to simplify the code:

```PHP
$last_element = end( $carry );
if ( isset( $last_element[0]['blockName'] ) && 'core/template-part' !== $last_element[0]['blockName'] ) {
	array_push( $last_element, $block );
	return $carry;
}
```

gigitux replied:
Mmm, I tried with your suggestion but the test failed. I suspect it happens because we don't push the updated element in the carry.

So, I prefer to not change anything here. What do you think?

Aljullu replied:
It sounds good, I might have missed something. If you keep it as-is, I think you'll need to add an extra check in the case that `count( $carry )` is 0, because then the code will try to use `$carry[ -1 ]` and produce a warning.

Aljullu commented on src/Templates/BlockTemplatesCompatibility.php:
```diff
@@ -361,4 +361,134 @@ protected function inject_attribute( &$block ) {
+		$parsed_blocks  = parse_blocks( $template_content );
+		$grouped_blocks = self::group_blocks( $parsed_blocks );
+
+		$single_product_template_blocks = array( 'woocommerce/breadcrumbs', 'woocommerce/product-gallery-image', 'woocommerce/product-add-to-cart', 'woocommerce/product-details', 'woocommerce/add-to-cart-form' );
```
When using the classic template, the breadcrumbs are shown outside of the `.product` div, so I think we should exclude `breadcrumbs` from this list.

![imatge](https://user-images.githubusercontent.com/3616980/218486401-ed485072-3f9e-456f-9a69-9e1ed6cc444f.png)

Also `woocommerce/product-add-to-cart` is used in the Products block, but not in the Single Product template, no? Should we remove it from the list as well? :thinking:

Aljullu reviewed (approved):
I left some other questions and comments regarding to the code. But once they are fixed/answered, it should be ready to go! Awesome work, @gigitux!

gigitux committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into fix/8314-blockifying-single-product-template-investigate-how-to-apply-wc-core-styles-to-product-template

gigitux committed:
address feedback

gigitux committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into fix/8314-blockifying-single-product-template-investigate-how-to-apply-wc-core-styles-to-product-template

gigitux committed:
Merge branch 'trunk' into fix/8314-blockifying-single-product-template-investigate-how-to-apply-wc-core-styles-to-product-template

gigitux committed:
fix name block and fix check

gigitux committed:
wrap single product template only on the frontend

gigitux commented:
Via [57cea7c](https://github.com/woocommerce/woocommerce-blocks/pull/8364/commits/57cea7cd4993fa6cfff5f4140233aa42e24fc3d3), I added another check to be sure that we wrap the template only on the frontend side. I will wait for your feedback before merging this PR! @Aljullu

Aljullu reviewed (approved):
Good catch. It makes sense to me.
